velox/functions/sparksql/Mask.cpp

```diff
@@ -0,0 +1,266 @@
+            digitChar,
+            otherChar,
+            row,
+            proxy);
+        proxy.finalize();
+      });
+    } else {
+      // The rest of the cases are handled through this general path and no
+      // direct access.
+      rows.applyToSelected([&](vector_size_t row) {
+        if (args[0]->isNullAt(row)) {
+          results->setNull(row, true);
+          return;
+        }
+        auto proxy = exec::StringWriter<>(results, row);
+        const auto upperChar =
+            getMaskedChar(upperChars, args[1], row, maskedUpperCase);
+        const auto lowerChar =
+            getMaskedChar(lowerChars, args[2], row, maskedLowerCase);
+        const auto digitChar =
+            getMaskedChar(digitChars, args[3], row, maskedDigit);
+        const auto otherChar =
+            (otherChars == nullptr || args[4]->containsNullAt(row))
+            ? std::nullopt
+            : std::optional(otherChars->valueAt<StringView>(row));
+        applyInner(
+            strings->valueAt<StringView>(row),
+            upperChar,
+            lowerChar,
+            digitChar,
+            otherChar,
+            row,
+            proxy);
+        proxy.finalize();
+      });
+    }
+  }
+
+  void applyInner(
+      StringView input,
+      const std::optional<StringView> upperChar,
+      const std::optional<StringView> lowerChar,
+      const std::optional<StringView> digitChar,
+      const std::optional<StringView> otherChar,
+      vector_size_t row,
+      facebook::velox::exec::StringWriter<false>& result) const {
+    const auto inputSize = input.size();
+    auto inputBuffer = input.data();
+    result.reserve(inputSize);
+    auto outputBuffer = result.data();
+
+    auto hasMaskedUpperChar = false;
+    auto hasMaskedLowerChar = false;
+    auto hasMaskedDigitChar = false;
+    auto hasMaskedOtherChar = false;
+    auto maskedUpperChar = "";
+    auto maskedLowerChar = "";
+    auto maskedDigitChar = "";
+    auto maskedOtherChar = "";
+
+    auto processChar = [&](const std::optional<StringView>& charOpt,
+                           const char*& maskedChar,
+                           bool& hasMaskedChar,
+                           const char* charType) {
+      if (charOpt.has_value()) {
+        VELOX_USER_CHECK(
+            charOpt.value().size() == 1,
+            std::string("Length of ") + charType + " should be 1");
+        maskedChar = charOpt.value().data();
+        hasMaskedChar = true;
+      }
+    };
+
+    processChar(upperChar, maskedUpperChar, hasMaskedUpperChar, "upperChar");
+    processChar(lowerChar, maskedLowerChar, hasMaskedLowerChar, "lowerChar");
+    processChar(digitChar, maskedDigitChar, hasMaskedDigitChar, "digitChar");
+    processChar(otherChar, maskedOtherChar, hasMaskedOtherChar, "otherChar");
+
+    for (auto i = 0; i < inputSize; i++) {
+      unsigned char p = inputBuffer[i];
+      if (isupper(p)) {
+        outputBuffer[i] = hasMaskedUpperChar ? maskedUpperChar[0] : p;
```

Comment:
We don't need processChar and hasMaskedUpperChar if the function inputs are std::optional\<char\> and can remove L169 - L194.

```
    for (auto i = 0; i < inputSize; i++) {
      unsigned char input = inputBuffer[i];
      unsigned char masked = input;
      if (isupper(p) && upperChar.has_value()) {
        masked = upperChar.value();
      } else if (islower(p) && lowerChar.has_value()) {
        masked = lowerChar.value();
      } else if (isdigit(p) && digitChar.has_value()) {
        masked = digitChar.value();
      } else if (otherChar.has_value()) {
        masked = otherChar.value();
      }
      outputBuffer[i] = masked;
    }
```